zerver/lib/webhooks/common.py

```diff
@@ -114,3 +115,39 @@ def validate_extract_webhook_http_header(request: HttpRequest, header: str,
         raise MissingHTTPEventHeader(header)
 
     return extracted_header
+
+
+def determine_http_headers(integration_name: str, fixture_name: str) -> Dict["str", "str"]:
+    """ In a nutshell, this method will try to find the headers corresponding to a particular
+    fixture for a particular integration. If a headers.py file exists for a specified integration,
+    it would mean that that particular integration has some fixtures that require extra HTTP headers
+    to operate and we should determine them and make them available during testing.
+
+    Note: The fixtures_to_headers_map does NOT need to contain the name of every fixture. It only
+    needs to contain the names of the fixtures which require http headers."""
+
+    headers_module_name = "zerver.webhooks.{integration_name}.headers".format(
+        integration_name=integration_name)
+    try:
+        # Try to import the headers.py file. If it does not exist, then that means that the
+        # integration does not have any special http headers that are required.
+        headers_module = importlib.import_module(headers_module_name)
+    except ImportError:
+        return {}
+    try:
+        fixtures_to_headers_map = headers_module.fixtures_to_headers_map  # type: ignore # we do extra exception handling in case it does not exist.
+    except AttributeError:
+        msg = " The {module_name} module must contain a fixtures_to_headers_map dictionary.".format(
+            module_name=headers_module_name
+        )
+        raise AttributeError(msg)
+    handler = fixtures_to_headers_map.get(fixture_name, None)
+    # This "handler" may either be a dictionary or a callable which would generate a
+    # dictionary. The reason why we allow callables is that sometimes, extra steps might
+    # have to be taken to generate the HTTP Headers, such as contacting a third party API
+    # endpoint first.
+    if not handler:
+        return {}
+    elif type(handler) == dict:
+        return handler
+    return handler()
```

Comment:
See, I am having trouble seeing how this makes things easier or more refined. 
* The user doesn't have to worry about headers at any given point. This is mostly for developer convenience.
* If we are the ones writing the integration, to begin with, I don't see how this makes life any easier. We already have a helper that can extract the header we want and we can set custom headers by passing them to `send_and_test_stream_message`. That seems simple (or complex) enough to me!
* Adding a separate module for headers seem overkill for passing in one custom value, and that too for just testing different fixtures. We don't need this extra layer of complexity.

Reply:
> This is mostly for developer convenience.

That's exactly what this is for :). It's for improving our integrations testing framework.

> We already have a helper that can extract the header we want and we can set custom headers by passing them to send_and_test_stream_message.

The thing is, we don't currently have a refined system for storing the headers, like we do with fixtures, for integrations. Our current system is directly passing them to `send_and_test_stream_message` in the tests. So that's the gap I'm trying to fill here. Also, it's needed if we want to pre-populate the headers fields in the integrations dev panel.

Also, which helper are you mentioning? I found a method in zerver/lib/webhooks/common called `validate_extract_webhook_http_header`, but that's for a totally different purpose (it's for use in writing the actual view code for the integration).

> Adding a separate module for headers seem overkill for passing in one custom value, and that too for just testing different fixtures

Perhaps, but there is the possibility that we might end up dealing with more than one custom value in the future. But otherwise, in some ways, I **do** agree with your perspective that this is added complexity. But I don't think that it's _too_ much extra complexity since only integrations which require HTTP headers would require the extra module/file.